pilot/pkg/xds/ads.go

```diff
@@ -793,30 +793,36 @@ func (s *DiscoveryServer) removeCon(conID string) {
 
 // Send with timeout
 func (conn *Connection) send(res *discovery.DiscoveryResponse) error {
-	done := make(chan error, 1)
+	errChan := make(chan error, 1)
 	// hardcoded for now - not sure if we need a setting
 	t := time.NewTimer(sendTimeout)
 	go func() {
-		err := conn.stream.Send(res)
-		conn.proxy.Lock()
-		if res.Nonce != "" {
-			if conn.proxy.Active[res.TypeUrl] == nil {
-				conn.proxy.Active[res.TypeUrl] = &model.WatchedResource{TypeUrl: res.TypeUrl}
-			}
-			conn.proxy.Active[res.TypeUrl].NonceSent = res.Nonce
-			conn.proxy.Active[res.TypeUrl].VersionSent = res.VersionInfo
-		}
-		conn.proxy.Unlock()
-		done <- err
+		errChan <- conn.stream.Send(res)
```

Comment:
is it normal to write to then close a channel? I will research

Reply:
sure. I was looking at some gRPC issue and stumbled on this comment https://github.com/grpc/grpc-go/issues/1229#issuecomment-302755717 as recommendation from gRPC team. So I compared our implementation and changed accordingly.

But sure good to research.

Reply:
https://play.golang.org/p/bMBjq8VbegF

Reply:
https://dave.cheney.net/2014/03/19/channel-axioms